### ISSUE TITLE
-fsanitize=signed-integer-overflow causes linker errors

### ISSUE THREAD

**Issue by ngg** [android-ndk/ndk#184]
If I compile with `-fsanitize=signed-integer-overflow` flag the binaries try to call the `__mulodi4` function and it will give an undefined reference linker error. This happens both on ARM and x86 archs. Most of the other similar sanitizers work well.

**Comment by stephenhines** [comment]
https://llvm.org/bugs/show_bug.cgi?id=28629 is the upstream bug for this. The truth is that we want to switch to compiler-rt, but just haven't had the ability to complete the transition. It is something we are looking at doing, however I don't have a great estimate on when it will be completed. You can add your own weak versions of those symbols in the meantime to get unblocked.

**Comment by sitsofe** [comment]
(For others who might end up here via a search: if you're trying to compile using clang's undefined sanitizer on a regular 32 bit Linux distro and you're hitting the "undefined reference to `__mulodi4'" a workaround might be to link against the appropriate libclang_rt.builtins* :

```shell
$ cat <<EOF > mulodi4.c 
int main(void) {
	long long ll = 1;
	int i1 = 2;
	return (ll * i1);
}
EOF
$ clang-4.0 -fsanitize=undefined mulodi4.c
/tmp/foo-605e44.o: In function `main':
foo.c:(.text+0x5a): undefined reference to `__mulodi4'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
$ clang-4.0 -fsanitize=undefined mulodi4.c /usr/lib/llvm-4.0/lib/clang/4.0.0/lib/linux/libclang_rt.builtins-i686.a 
$ echo $?
0
```

See https://github.com/ldc-developers/ldc/issues/813 for an example of a project that works around the problem)

**Comment by DanAlbert** [comment]
This no longer happens for this particular case in r17. Seems Clang isn't generating this call for signed-integer-overflow any more.

Closing in favor of https://github.com/android-ndk/ndk/issues/294 as the root cause rather than the symptom.

**Comment by AstralStorm** [comment]
Happens again in r18.
